:white_check_mark: Add frontend security with a login page

JonasZeihe opened the pull request:
you can now login to the app, get an accestoken from backend and view protected routes

frontend security funktioniert, man kann sich mit einem Nutzernamen und Passwort einloggen und so auf protected routes zugreifen

JonasZeihe committed:
:white_check_mark: Add frontend security with a login page

jamarob commented on frontend/src/App.js:
```diff
@@ -1,17 +1,25 @@
-    </Switch>
+    <AuthProvider>
+      <Switch>
+        <Route exact path="/login">
+          <Login />
+        </Route>
+        <ProtectedRoute exact path="/home">
+          <CourseOverview />
+        </ProtectedRoute>
+        <ProtectedRoute exact path="/course/:courseName">
+          <CourseDetails />
+        </ProtectedRoute>
```
`exact` is not needed here. 

Btw, what do you think about 
```javascript
<Route path="/home" component={Home} />
```
short and sweet :)

jamarob commented on frontend/src/auth/AuthenticationContext.js:
```diff
@@ -0,0 +1,5 @@
+import { useContext, createContext } from 'react'
```
This file is a duplicate of `AuthContext` can be deleted, because it is not used.

jamarob commented on frontend/src/auth/AuthenticationProvider.js:
```diff
@@ -0,0 +1,11 @@
+import { useState } from 'react'
```
Duplicate of `AuthProvider`

jamarob commented on frontend/src/pages/CourseDetails.js:
```diff
@@ -1,22 +1,24 @@
   useEffect(() => {
-    getCourseByName(courseName).then(setCourseData)
-    getUsersByCourseName(courseName)
+    getCourseByName(courseName, token).then(setCourseData)
```
Error handling?

jamarob commented on frontend/src/pages/CourseDetails.js:
```diff
@@ -38,8 +40,8 @@ export default function CourseDetails() {
   const deleteUser = (userId) => {
-    deleteUserById(userId).then(() => {
-      setUsers(users.filter((user) => user.name !== userId))
+    deleteUserByName(userId).then(() => {
```
Error handling?

jamarob commented on frontend/src/pages/Login.js:
```diff
@@ -0,0 +1,46 @@
+
+  const handleSubmit = (event) => {
+    event.preventDefault()
+    if (!userName && !userPassword) {
```
```suggestion
    if (!(userName && userPassword)) {
```

or
```suggestion
    if (!userName || !userPassword) {
```

jamarob commented on frontend/src/pages/Login.js:
```diff
@@ -0,0 +1,46 @@
+    if (!userName && !userPassword) {
+      return
+    }
+    loginUser(userName, userPassword).then(setToken)
```
Error handling!

jamarob commented on frontend/src/pages/Login.js:
```diff
@@ -0,0 +1,46 @@
+          placeholder="Username"
+          type="text"
+          value={userName}
+          onChange={({ target }) => setUserName(target.value)}
```
Good use of destructuring!

jamarob commented on frontend/src/services/apiService.js:
```diff
@@ -1,24 +1,33 @@
-import axios from 'axios'
+import axiosConfig from './axiosConfig'
```
Maybe 
```suggestion
import {axiosInstance as axios} from './axiosConfig'
```

jamarob commented on frontend/src/services/axiosConfig.js:
```diff
@@ -0,0 +1,12 @@
+import axios from 'axios'
+
+const axiosInstance = axios.create()
+
+function setAxiosAuthToken(token) {
+  axiosInstance.defaults.headers.common.Authorization = 'Bearer ' + token
+}
+
+export default {
+  axiosInstance,
+  setAxiosAuthToken,
+}
```
I prefer this way
```suggestion
export const axiosInstance = axios.create()

export function setAxiosAuthToken(token) {
  axiosInstance.defaults.headers.common.Authorization = 'Bearer ' + token
}
```

but thats really a matter of taste

jamarob reviewed (commented):
Good job 👍 Some places need error handling or your app will crash in case of a network error.

JonasZeihe committed:
:bug: Fixed deleteUserByName now it is deleteUserByCourseNameAndName

now you can create courses, create users with the same name but different courses, and delete every user in a specific course without deleting users with the same name in other courses
